### PR TITLE
[MAINTENANCE] Refactor Anonymizer utilizing the Strategy design pattern

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Refactor Anonymizer using the Strategy design pattern to create a single, uniform interface for callers to interact with.
  - This abstracts away the need to know about the child classes.

Here are the following layers to be aware of:
* `BaseAnonymizer`
  * Contains generic behavior that applies to all or most `Anonymizer` child classes. Additionally, contains abstract methods `anonymize` and `can_handle` that are to be leveraged by children to conditionally anonymize input objects.
* `ProfilerRunAnonymizer`/`CheckpointRunAnonymizer`/etc
  * Fulfills the contract set by `BaseAnonymizer` by inheriting the class and fleshing out the aforementioned methods.
  * Contain a reference to the primary `Anonymizer` - see below for what that does!
* `Anonymizer`
  * Wrapper class that conducts the "strategy"
  * Iterates through possible specific `Anonymizer` classes, sees if one `can_handle`, and conditionally determines which "strategy" to use for anonymization.
  * In the case a specific anonymizer class needs behavior it isn't responsible for (i.e. `ProfilerAnonymizer` needs an anonymized batch request so it has to talk to `Anonymizer` to get those results).

This allows us to make all anonymizer behavior private and encapsulated from external users. If you want to anonymize something, you need to call `Anonymizer.anonymize()`.

The diff is HUGE but that's mainly because there are a lot of moving parts. Let me detail what's important to review:
* base.py - The abstract class with shared behavior
* anonymizer.py - The main class that conducts the "strategy"
  - Note that this implementation has been impacted by a whole lot of cyclic import issues. I've made design decisions that I wouldn't otherwise have made (using instance attributes when class attributes would make a lot more sense).
 
Outside of that, everything is a specific object anonymizer implementation (which has existed in the past) or changes to existing code to use this new, uniform API (`Anonymizer.anonymize()`).

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/core/usage_statistics/anonymizers/base.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/base.py:14]: Take all shared behavior from `Anonymizer` and place here - this is subclassed throughout. All child classes must fulfill the two abstract methods.

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: This is the core logic - check all each anonymizer strategy in `self.strategies`.

If that strategy `can_handle`, use that implementation's `anonymize` method. If no specialized `Anonymizer` child class works, we revert to generic strategies as noted in `self._anonymize`.

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: Base class can handle everything!

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: Move logic from `usage_statistics` to `anonymizer` - most of the method is anonymizing behavior so it makes more sense here. Additionally, this lets us keep all methods private.

Comment (anthonyburdi) [reply great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: Totally works - this feels a bit weird to me though from a responsibility standpoint to include in any of the Anonymizer classes. I don't have a better idea though so just thinking out loud :)

File: great_expectations/core/usage_statistics/anonymizers/checkpoint_run_anonymizer.py
Comment (anthonyburdi) [inline great_expectations/core/usage_statistics/anonymizers/checkpoint_run_anonymizer.py]: Just an optional suggestion - could this be `_anonymize_checkpoint_config()` just to be more descriptive that it acts on checkpoint config dicts?

File: great_expectations/core/usage_statistics/usage_statistics.py
Comment (anthonyburdi) [inline great_expectations/core/usage_statistics/usage_statistics.py]: Looks like you can remove these 3 comments

File: great_expectations/data_context/types/base.py
Comment (anthonyburdi) [inline great_expectations/data_context/types/base.py:2575]: 👏 👏 👏 👏 👏  I love this method here :)

**Review by anthonyburdi — COMMENTED** [review]
LGTM! I'll hold off on approving so @donaldheppner can take a look.

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (cdkini) [reply great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: Ah yes this is a very valid point. To be perfectly honest, the primary purpose here is to prevent having to expose functionality from the anonymizer.

By keeping the logic here, we can keep all helpers private and maintain the ideal that `anonymize` is the single public entrypoint into anonymizer behavior.

If you feel strongly about this (and I think that's totally justifiable), let me know and I can figure something out.

File: great_expectations/core/usage_statistics/usage_statistics.py
Comment (cdkini) [reply great_expectations/core/usage_statistics/usage_statistics.py]: done! great catch 🙇🏽

File: great_expectations/core/usage_statistics/anonymizers/checkpoint_run_anonymizer.py
Comment (cdkini) [reply great_expectations/core/usage_statistics/anonymizers/checkpoint_run_anonymizer.py]: I agree - done!

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: We have a few problems that this implementation solves:
1. Cyclic imports - Our "strategies" should definitely be class level attributes but there's no feasible way to make that work with the current state of the codebase. I've opted to place them in the constructor here to shield them from the global scope, allowing us to get by without issues.
2. Infinite recursion - if I were to instantiate all these child anonymizers right here and then use an instance of this anonymizer in the child classes, there would be infinite recursion.

Let me try to explain this as best as I can. Our primary `Anonymizer` class is sort of the web between all anonymizers in GE. While this class references specific implementations, those specific implementations also reference back here. This is deemed necessary because if a particular class requires behavior that is outside the responsibility of its class, it requests that behavior from this primary `Anonymizer` class. 

If my `DatasourceAnonymizer` has a constructor like this:
```python
def __init__(self, salt, aggregate_anonymizer):
    ...
```
it becomes evident that part of the definition of the class is its relationship with the primary `Anonymizer`. This bidrectional link causes recursion issues because both classes are defined in terms of each other. Instead, I've opted to only instantiate what we need at a given time and cache those results. This could probably be explained better live but this solution here seems to have mitigated much of the pain of this refactor.

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/anonymizer.py]: 1. Look through our strategies and see if we have a solution for the problem at hand
2. If we do, look at the cache and try to get an already instantiated object to do the work for us
  - If we have a cache miss, create a new instance and add to the cache
3. Run the anonymize method on the instance that's responsible for the object we're currently working with

File: great_expectations/core/usage_statistics/anonymizers/base.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/base.py]: This is still the core part of this class - what does a class need to be able to do to be considered an anonymizer?

Note that all the other methods (outside a few key ones) have been refactored out into separate anonymizer classes like before.

File: great_expectations/core/usage_statistics/anonymizers/batch_anonymizer.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/anonymizers/batch_anonymizer.py]: Let's dissect this a bit:

We take in `salt` and an `aggregate_anonymizer`.

The former is also defined in the base class and is used to offset hashes when anonymizing payloads. The latter is a reference to the primary `Anonymizer` class - the actual `Anonymizer` will generally pass itself here so we can traverse back and forth from `BatchAnonymizer` to `Anonymizer` (and vice-versa).

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (donaldheppner) [inline great_expectations/core/usage_statistics/anonymizers/anonymizer.py:51]: I think you can make these instances and replace `CheckpointAnonymizer` with `CheckpointAnonymizer(self)`

File: great_expectations/core/usage_statistics/anonymizers/action_anonymizer.py
Comment (donaldheppner) [inline great_expectations/core/usage_statistics/anonymizers/action_anonymizer.py]: Do we need this reference to be optional?

File: great_expectations/core/usage_statistics/anonymizers/anonymizer.py
Comment (donaldheppner) [reply great_expectations/core/usage_statistics/anonymizers/anonymizer.py:51]: This would also allow can_handle to be an instance method

**Review by donaldheppner — CHANGES_REQUESTED** [review]

**Review by donaldheppner — APPROVED** [review]

**Pull request merged**
